src/client/dfs/dfs.c

```diff
@@ -1677,6 +1718,15 @@ dfs_remove(dfs_t *dfs, dfs_obj_t *parent, const char *name, bool force,
 	if (rc)
 		return rc;
 
+	if (dfs->use_dtx) {
+		rc = daos_tx_open(dfs->coh, &th, 0, NULL);
+		if (rc) {
+			D_ERROR("daos_tx_open() failed (%d)\n", rc);
+			D_GOTO(out, rc);
+		}
+	}
+
+restart:
 	/** Even with cond punch, need to fetch the entry to check the type */
 	rc = fetch_entry(parent->oh, th, name, false, &exists, &entry);
 	if (rc)
```

Comment:
A read may also return -DER_TX_RESTART, when its epoch is potentially a bit low---failing the epoch uncertainty check. What would be the best way to structure the code to handle it is a question... (E.g., perhaps go to the error handling after calling daos_tx_commit might be an option.)

Reply:
hmm, yes that's a bit challenging because here we have converted the error code to errno. i would need to update this to return the original restart error to make sure we catch that and restart the tx.

Reply:
hmm this is actually not just for fetch i guess, but also enumerate.
this might be a pain to handle in the current code.

Reply:
Yes, this burden applies to all TX reads (i.e., fetch, list, query), TX readwrites (i.e., conditional), and daos_tx_commit. I'm just thinking out aloud here: Would a daos_tx_error help here?

    rc = <read or readwrite>(...);
    if (rc is not a nonerror like EXIST, NONEXIST)
        goto out;
    daos_tx_commit(...);
    out:
    rc = daos_tx_error(th);
    if (rc == -DER_TX_RESTART)
        ...
    else
        ...

(I view this PR as a test of our TX API design. The design might need to be improved.)

Reply:
In other words, would it be easier if the application code only decides "should I continue executing the TX or should I stop":

- TX error EXIST/NONEXIST: app may decide to continue executing the TX;
- TX error like TX_RESTART: app should stop;
- App error like NOMEM: app should stop;

When the TX stops:

- If daos_tx_error(th) != 0, handle TX_RESTART;
- Otherwise, handle app error.